pkg/ruler/ruler.go

```diff
@@ -518,6 +562,70 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 	return ownsRuleGroup, nil
 }
 
+func (r *Ruler) LivenessCheck(_ context.Context, request *LivenessCheckRequest) (*LivenessCheckResponse, error) {
+	level.Debug(r.logger).Log("msg", "liveness check request", "request received from", request.RulerAddress)
+	if r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped() {
+		return nil, errors.New("ruler's context is canceled and might be stopping soon")
+	}
+	if !r.subservices.IsHealthy() {
+		return nil, errors.New("not all subservices are in healthy state")
+	}
+	return &LivenessCheckResponse{State: int32(r.State())}, nil
+}
+
+// This function performs a liveness check against the provided replicas. If any one of the replicas responds with a state = Running, then
+// this Ruler should not take ownership of the rule group. Otherwise, this Ruler must take ownership of the rule group to avoid missing evaluations
+func nonPrimaryInstanceOwnsRuleGroup(r ruler, g *rulespb.RuleGroupDesc, replicas []string, selfAddress string) bool {
+	userID := g.User
+
+	rgToken := tokenForGroup(g)
+	jobs := concurrency.CreateJobsFromStrings(replicas)
+
+	errorChan := make(chan error, len(jobs))
+	responseChan := make(chan *LivenessCheckResponse, len(jobs))
+
+	ctx := user.InjectOrgID(context.Background(), userID)
+	ctx, cancel := context.WithTimeout(ctx, r.Config().LivenessCheckTimeout)
+	defer cancel()
+
+	err := concurrency.ForEach(ctx, jobs, len(jobs), func(ctx context.Context, job interface{}) error {
+		addr := job.(string)
+		rulerClient, err := r.GetClientFor(addr)
+		if err != nil {
+			errorChan <- err
+			level.Debug(r.Logger()).Log("msg", "unable to get client for ruler", "ruler addr", addr, "token", rgToken)
+			return nil
+		}
+		level.Debug(r.Logger()).Log("msg", "performing liveness check against", "addr", addr, "for", g.Name, "token", rgToken, "instance addr", selfAddress)
+
+		resp, err := rulerClient.LivenessCheck(ctx, &LivenessCheckRequest{
+			RulerAddress: selfAddress,
+		})
+		if err != nil {
+			errorChan <- err
+			level.Debug(r.Logger()).Log("msg", "liveness check failed", "addr", addr, "for", g.Name, "err", err.Error(), "token", rgToken)
+			return nil
+		}
+		level.Debug(r.Logger()).Log("msg", "liveness check succeeded ", "addr", addr, "for", g.Name, "token", rgToken, "ruler state", services.State(resp.GetState()))
+		responseChan <- resp
+		return nil
+	})
+
+	close(errorChan)
+	close(responseChan)
```

Comment:
if one ruler is taking time to reply the liveness, wouldn't it make this whole call to hang?

Reply:
discussed offline... i just think the timeout could be hardcoded and not a config.

Reply:
Set the timeout to 100ms and removed the config